hazelcast/src/main/java/com/hazelcast/internal/networking/nio/NioOutboundPipeline.java

```diff
@@ -208,7 +208,8 @@ public void write(OutboundFrame frame) {
                     }
                 } else {
                     // no write through, let the io thread deal with it.
-                    ownerAddTaskAndWakeup(this);
+                    registerOp(OP_WRITE);
```

Comment:
This is the magical part. So instead of adding the pipeline to the taskQueue of the owner and then waking it up, we register for an OP_WRITE and then we call selector.wakeup.

Reply:
@pveentjer One potential reason why we may observe worse latencies is that registration of selection key is blocking operation. Typically, all calls to selector methods are wrapped in a synchronized block. So registration of a new selection key could be treated as a call to blocking queue. E.g. this is what happens in Linux in JDK 11 [1]. Before the change we did concurrent queue + wakeup, now we do blocking queue + wakeup.

Most probably we need to have some throttling mechanism, which will prevent a call to the selector if the write operation is already scheduled. E.g. AtomicBoolean.

WDYT?

[1] https://github.com/openjdk/jdk11u/blob/737d8437886ad97c6ed21a25b9911c10b3886f61/src/java.base/linux/classes/sun/nio/ch/EPollSelectorImpl.java#L242

Reply:
This scheduling logic is already in place. So if there are 10 simultaneous writes, only 1 write will schedule the pipeline and that thread will update the selectionkey.interestset. This is all done under the protection of the scheduled flag.